src/calendar/calendar.tsx

```diff
@@ -0,0 +1,394 @@
+            case "Year":
+                if (showView === "Month") {
+                    if (goTo.length) setCalendarDate(dayjs(goTo));
+                    setShowView("Day");
+                    break;
+                }
+
+                if (showView === "Year") {
+                    if (goTo.length) setCalendarDate(dayjs(goTo));
+                    setShowView("Day");
+                    return;
+                }
+
+                setShowView("Year");
+                break;
+            default:
+                setShowView("Day");
+                break;
+        }
+    };
+
+    const handleDayClick = (value: string) => {
+        const isNextMonthSelected =
+            calendarDate?.clone().toDate().getMonth() !== dayjs(value).month();
+
+        if (isNextMonthSelected) setCalendarDate(dayjs(value));
+
+        setSelectedStartDate(value);
+
+        performOnChangeHandler(value);
+    };
+
+    const handleMonthYearClick = (
+        e: React.MouseEvent<HTMLDivElement>,
+        show: View
+    ) => {
+        const value = e.currentTarget.getAttribute("data-value") || "";
+        let isYearDecaded = false;
+        if (show === "Year") {
+            isYearDecaded =
+                e.currentTarget.getAttribute("data-variant") === "nextDecaded"
+                    ? true
+                    : false;
+        }
+
+        if (!value.length) return;
+        const hasStartDate: string[] = selectedStartDate.split("-");
+
+        const [yyyy, mm, _dd] = hasStartDate;
+        const selectedMonth = value.split("-")[1];
+        const selectedYear = value.split("-")[0];
+
+        if (hasStartDate.length === 3) {
+            /**
+             * Handle last day check/validate
+             * user select 2023-01-30 and select month calendar which didn't exist 30 day in the month
+             * it will get the last day of the month
+             */
+            const lastDayInMonth: number = dayjs(
+                `${yyyy}-${selectedMonth}-01`
+            ).daysInMonth();
+            let dd: string = _dd;
+
+            if (+_dd > lastDayInMonth) {
+                dd = lastDayInMonth.toString();
+            }
+
+            // date already been selected
+            const fullDate =
+                show === "Month"
+                    ? `${selectedYear}-${selectedMonth}-${dd}`
+                    : showView === "Year"
+                    ? `${selectedYear}-${mm}-${dd}`
+                    : dayjs().format("YYYY-MM-DD");
+
+            if (isYearDecaded) setCalendarDate(dayjs(value));
+            setSelectedStartDate(fullDate);
+            return;
+        }
+
+        // wihtout selected date yet
+        if (isYearDecaded) setCalendarDate(dayjs(value));
+        setSelectedStartDate(value);
+    };
+
+    // =============================================================================
+    // Calendar Generate
+    // =============================================================================
```

Comment:
The memoise hook use for prevent it do render in unnecessary update which may be the state been updated in other component which cause this calendar re-calculated again. correct me if i'm wrong to use the useMemo.

Reply:
Question:

The calendar will be generate depend on the `calendarDate`. You need a hook for the useCalendar.tsx file or ts file?

```
const firstDayOfTheMonth = useMemo(
    () => calendarDate.clone().startOf("month"),
    [calendarDate]
);

```

Reply:
it's a good idea to memoise the final calculated value. but since `firstDayOfTheMonth` and `firstDayOfFirstWeekOfMonth` are both derived from `calendarDate`, they can all be computed together

I'm proposing

```
// src/utils/calendar-helper.ts

const generateWeeksOfTheMonth = (calendarDay: Dayjs) => {
  const firstDayOfTheMonth = ...
  const firstDayOfFirstWeekOfMonth = ...
  const firstDayOfEachWeek = ...
  return firstDayOfEachWeek.map(...)
}

// src/calendar/calendar.tsx

const weeksOfTheMonth = useMemo(() => generateWeeksOfTheMonth(calendarDay), [calendarDay]);
```

can correct me if I missed out any dependencies

Reply:
Wa. you response the final answer. Thanks to guide me. I learn bit from you. 

I was thinking that was a useHook instead of ts file.